libdnf5/conf/vars.cpp

```diff
@@ -180,67 +182,243 @@ std::unique_ptr<std::string> Vars::detect_release(const BaseWeakPtr & base, cons
+                        continue;
+                    }
+                    pos_after_variable_expression = pos_after_word + 1;
+                } else if (res[pos_after_variable] == '}') {
+                    // ${variable}
+                    if (variable_mapping != variables.end()) {
+                        subst_str = &variable_mapping->second.value;
+                    }
+                    // Move past the closing '}'
+                    pos_after_variable_expression = pos_after_variable + 1;
+                } else {
+                    // Variable expression doesn't end in a '}', continue after the variable
+                    pos = pos_after_variable;
+                    continue;
+                }
+            } else {
+                // No braces, we have a $variable
+                if (variable_mapping != variables.end()) {
+                    subst_str = &variable_mapping->second.value;
+                }
+                pos_after_variable_expression = pos_after_variable;
+            }
+
+            // If there is no substitution to make, move past the variable expression and continue.
+            if (subst_str == nullptr) {
+                total_scanned += pos_after_variable_expression - pos;
+                pos = pos_after_variable_expression;
+                continue;
+            }
+
+            res.replace(pos, pos_after_variable_expression - pos, *subst_str);
+            total_scanned += pos_after_variable_expression - pos;
+            pos += subst_str->length();
         } else {
-            start = res.find_first_of('$', past_variable);
+            total_scanned += 1;
+            pos += 1;
         }
     }
 
-    return res;
+    // We have reached the end of the text
+    if (depth > 0) {
+        // If we are in a subexpression and we didn't find a closing '}', make no substitutions.
+        return std::make_pair(std::string{text}, text.length());
+    }
+
+    return std::make_pair(res, text.length());
+}
+
+std::string Vars::substitute(const std::string & text) const {
+    return substitute_expression(text, 0).first;
+}
+
+std::tuple<std::string, std::string> Vars::split_releasever(const std::string & releasever) {
+    // Uses the same logic as splitReleaseverTo in libzypp
+    std::string releasever_major;
+    std::string releasever_minor;
+    const auto pos = releasever.find('.');
+    if (pos == std::string::npos) {
+        releasever_major = releasever;
+    } else {
+        releasever_major = releasever.substr(0, pos);
+        releasever_minor = releasever.substr(pos + 1);
+    }
+    return std::make_tuple(releasever_major, releasever_minor);
+}
+
+bool Vars::is_read_only(const std::string & name) {
+    return READ_ONLY_VARIABLES.contains(name);
 }
 
 void Vars::set(const std::string & name, const std::string & value, Priority prio) {
+    if (is_read_only(name)) {
+        throw RuntimeError(M_("Variable \"{}\" is read-only"), name);
```

Comment:
Please document that it can throw an error. I have to test, but I think that it will create a problem with setvars option or other channels that can modify variables.

Reply:
I've tested it and it works like expected. I am still not sure whether RuntimeError is the best exception.

Reply:
I know that it is used in the rest of the module but I feel that libdnf5 exception that inherits from runtime would be more appropriate. What do you think?

Reply:
I agree, I added a `ReadOnlyVariableError` class, and I put a `/// @throw ReadOnlyVariableError` in the docstring. Let me know if I should document it anywhere else.

I added a test here https://github.com/rpm-software-management/ci-dnf-stack/pull/1359 to ensure read-only variables cannot be set.

Reply:
Additionally I asked @jrohel about opinion for locking mechanism, because he is an author of config locking mechanism.

Reply:
This is not about locking. This code implements a hard-coded list of read-only variable names. The value of these variables is not locked, but is calculated based on the value of another variable. In other words, the value of these variables is not locked, they can be changed at any time, indirectly, by changing the value of another variable.